sql/2021/pwa/lighthouse_pwa_audits_sw.sql

```diff
@@ -0,0 +1,48 @@
+#standardSQL
+# Get summary of all lighthouse scores for a category for PWA pages (i.e. those with a service worker and a manifest file)
+# Note scores, weightings, groups and descriptions may be off in mixed months when new versions of Lighthouse roles out
+
+CREATE TEMPORARY FUNCTION getAudits(report STRING, category STRING)
+RETURNS ARRAY<STRUCT<id STRING, weight INT64, audit_group STRING, title STRING, description STRING, score INT64>> LANGUAGE js AS '''
+var $ = JSON.parse(report);
+var auditrefs = $.categories[category].auditRefs;
+var audits = $.audits;
+$ = null;
+var results = [];
+for (auditref of auditrefs) {
+  results.push({
+    id: auditref.id,
+    weight: auditref.weight,
+    audit_group: auditref.group,
+    description: audits[auditref.id].description,
+    score: audits[auditref.id].score
+  });
+}
+return results;
+''';
+
+SELECT
+  audits.id AS id,
+  COUNTIF(audits.score > 0) AS num_pages,
+  COUNT(0) AS total,
+  COUNTIF(audits.score IS NOT NULL) AS total_applicable,
+  SAFE_DIVIDE(COUNTIF(audits.score > 0), COUNTIF(audits.score IS NOT NULL)) AS pct,
+  APPROX_QUANTILES(audits.weight, 100)[OFFSET(50)] AS median_weight,
+  MAX(audits.audit_group) AS audit_group,
+  MAX(audits.description) AS description
+FROM
+  `httparchive.lighthouse.2021_07_01_mobile`
+JOIN
+  `httparchive.pages.2021_07_01_mobile`
+USING (url),
+  UNNEST(getAudits(report, "pwa")) AS audits
+WHERE
+  JSON_EXTRACT(payload, '$._pwa') != "[]" AND
+  JSON_EXTRACT(payload, '$._pwa.serviceWorkers') != "[]" AND
```

Comment:
@tunetheweb: leaving this comment here, so we can discuss if it makes sense to include `$._pwa.serviceWorkerHeuristic` in the final query.

Reply:
Yes, as discussed previously, we’ll make the call on using this for all the queries after a July has run and we see the full numbers.

Reply:
July has now finished so let's look at the stats:

Running Lighthouse we see:

id | num_pages | total | total_applicable | pct
-- | -- | -- | -- | --
service-worker | 107,775 | 7,262,217 | 7,165,208 | 1.50%

Running blink usage and our custom metrics we have:

client | sw_blink_usage | serviceworkers | serviceWorkerHeuristic | total | serviceworkers_pct | serviceWorkerHeuristic_pct | sw_blink_usage_pct
-- | -- | -- | -- | -- | -- | -- | --
desktop | 77,020 | 55,723 | 191,745 | 6,286,373 | 0.89% | 3.05% | 1.23%
mobile | 89,510 | 76,011 | 240,942 | 7,491,840 | 1.01% | 3.22% | 1.19%

So it appears Lighthouse over counts compared to blink usage and our custom stat, and the heuristic way over counts. Given that, it seems our original `serviceworkers` is a better metric to use rather than the `serviceWorkerHeuristic` megric even if it does under count blink usage by 0.34% on desktop and 0.18% on mobile.

Lighthouse over counts blink usage by 0.31%, and our `serviceworkers` metric by 0.49%.

Comparing this with [Last years stats](https://docs.google.com/spreadsheets/d/1AOqCkb5EggnE8BngpxvxGj_fCfssT9sZ0ElCQKp4pOw/edit#gid=814184412):

[Lighthouse](https://docs.google.com/spreadsheets/d/1AOqCkb5EggnE8BngpxvxGj_fCfssT9sZ0ElCQKp4pOw/edit#gid=814184412):

id | num_pages | total | total_applicable | pct
-- | -- | -- | -- | --
service-worker | 64,373 | 6,290,131 | 6,252,154 | 1.03%

[Blink usage](https://docs.google.com/spreadsheets/d/1AOqCkb5EggnE8BngpxvxGj_fCfssT9sZ0ElCQKp4pOw/edit#gid=2077755325):

client | feature | num_urls | total_urls | pct
-- | -- | -- | -- | --
desktop | ServiceWorkerControlledPage | 49,305 | 5,593,642 | 0.88%
mobile | ServiceWorkerControlledPage | 55,019 | 6,347,919 | 0.87%

[Almanac Custom Stats](https://docs.google.com/spreadsheets/d/1AOqCkb5EggnE8BngpxvxGj_fCfssT9sZ0ElCQKp4pOw/edit#gid=227354805):

date | client | serviceworkers_pct | total
-- | -- | -- | -- | -- | -- | --
2020-08-01 | desktop | 0.76% | 5,593,642
2020-08-01 | mobile | 0.84% | 6,347,640

We can see our custom stat also undercounted blink usage last year, though by less (0.12% on desktop, and 0.03% on mobile), and both undercounted Lighthouse even more (by 0.16% for blink usage and 0.19% for our stat).

This again points us at using the `serviceworkers` custom metric over `serviceWorkerHeuristic` as, although the divergence from blink usage and Lighthouse has got slightly worse, it's similar to what we used last year and the `serviceWorkerHeuristic` is just over counting way too much.

It's also notable that we have nearly a million extra URLs in this years data, compared to last year. Whether this is new sites, or just more of a long tail (who may or may not use Service Workers more or less) remains to be seen.

All of these stats show that we have different numbers for each, and probably none of these numbers are "correct", and the truth probably lies somewhere in between these all.

However I do think the numbers we have gathered with our `serviceworkers` metric are a significant proportion of the PWA sites out there, and so are likely to be representative of the usage of service workers and manifests in a similar proportion to they were last year, so while we can prefer blink usage data for reporting actual service worker usage (as we did last year), the custom metric data can still be used (as it was last year, though in a slightly different guise) for other data that we do not have other sources for (like manifest usage, popularity of SW events, workbox usage...etc.). Though, like last year, I would advise reporting in terms of percentages rather than absolute number of sites, since our data is not (and cannot be) 100% accurate to the exact usage. This is the preference within the Web Almanac anyway, to allow easier comparison year on year amongst other reasons.

Let me know if you agree and so can resolve this comment (and others like it) is we stick with `serviceworks` rather than switching to `serviceWorkerHeuristic`?

**Queries:**

Lighthouse:

```sql
#standardSQL
# Get summary of all lighthouse scores for a category
# Note scores, weightings, groups and descriptions may be off in mixed months when new versions of Lighthouse roles out

CREATE TEMPORARY FUNCTION getAudits(auditRefs STRING, audits STRING)
RETURNS ARRAY<STRUCT<id STRING, weight INT64, audit_group STRING, title STRING, description STRING, score INT64>> LANGUAGE js AS '''
var auditrefs = JSON.parse(auditRefs);
var audits = JSON.parse(audits);
var results = [];
for (auditref of auditrefs) {
  results.push({
    id: auditref.id,
    weight: auditref.weight,
    audit_group: auditref.group,
    description: audits[auditref.id].description,
    score: audits[auditref.id].score
  });
}
return results;
''';

SELECT
  audits.id AS id,
  COUNTIF(audits.score > 0) AS num_pages,
  COUNT(0) AS total,
  COUNTIF(audits.score IS NOT NULL) AS total_applicable,
  SAFE_DIVIDE(COUNTIF(audits.score > 0), COUNTIF(audits.score IS NOT NULL)) AS pct,
  APPROX_QUANTILES(audits.weight, 100)[OFFSET(50)] AS median_weight,
  MAX(audits.audit_group) AS audit_group,
  MAX(audits.description) AS description
FROM
  `httparchive.lighthouse.2021_07_01_mobile`,
  --`httparchive.sample_data.lighthouse_mobile_10k`,
  UNNEST(getAudits(JSON_EXTRACT(report, '$.categories.pwa.auditRefs'),JSON_EXTRACT(report, '$.audits'))) AS audits
GROUP BY
  audits.id
ORDER BY
  median_weight DESC,
  id
```

Blink usage and custom metrics:
```sql
#standardSQL
# Counting Manifests and Service Workers
SELECT
  client,
  SUM(ServiceWorkerControlledPage) as sw_blink_usage,
  SUM(ServiceWorker) as serviceworkers,
  SUM(serviceWorkerHeuristic) as serviceWorkerHeuristic,
  SUM(COUNT(0)) OVER (PARTITION BY client) AS total,
  SAFE_DIVIDE(SUM(ServiceWorker), SUM(COUNT(0)) OVER (PARTITION BY client)) AS serviceworkers_pct,
  SAFE_DIVIDE(SUM(serviceWorkerHeuristic), SUM(COUNT(0)) OVER (PARTITION BY client)) AS serviceWorkerHeuristic_pct,
  SAFE_DIVIDE(SUM(ServiceWorkerControlledPage), SUM(COUNT(0)) OVER (PARTITION BY client)) AS sw_blink_usage_pct
FROM (
  SELECT
    _TABLE_SUFFIX AS client,
    IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.serviceWorkers') != "[]",1,0) AS ServiceWorker,
    IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.serviceWorkerHeuristic') = 'true',1,0) AS serviceWorkerHeuristic,
    IF(IFNULL(
    JSON_EXTRACT(payload, '$._blinkFeatureFirstUsed.Features.ServiceWorkerControlledPage'),
    JSON_EXTRACT(payload, '$._blinkFeatureFirstUsed.Features.990'))
  IS NOT NULL, 1, 0) AS ServiceWorkerControlledPage,
  FROM
    --`httparchive.sample_data.pages_*`
    `httparchive.pages.2021_07_01_*`
  )
GROUP BY
  client
ORDER BY
  client
```

Reply:
Actually looking into this some more I'm not convinced.

I've created a scratch table with the following SQL:

```sql
#standardSQL
CREATE TABLE
`httparchive.scratchspace.serviceworkers_2021_07_01_mobile`
AS
SELECT
  url,
  JSON_EXTRACT(report, '$.audits.service-worker.score') as lh_service_worker,
  IF(IFNULL(
    JSON_EXTRACT(payload, '$._blinkFeatureFirstUsed.Features.ServiceWorkerControlledPage'),
    JSON_EXTRACT(payload, '$._blinkFeatureFirstUsed.Features.990'))
  IS NOT NULL, 1, 0) AS ServiceWorkerControlledPage,
  IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.serviceWorkers') != "[]",1,0) AS ServiceWorker,
  IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.workboxInfo') != "[]",1,0) AS workboxInfo,
  IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.swEventListenersInfo') != "[]",1,0) AS swEventListenersInfo,
  IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.swMethodsInfo') != "[]",1,0) AS swMethodsInfo,
  IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.serviceWorkerHeuristic') = 'true',1,0) AS serviceWorkerHeuristic,
  JSON_EXTRACT(payload, '$._pwa') as pwaCustomMetric
FROM
    `httparchive.pages.2021_07_01_mobile`
JOIN
    `httparchive.lighthouse.2021_07_01_mobile`
USING
  (url)
WHERE
  (JSON_EXTRACT(payload, '$._pwa') != "[]" AND
    (JSON_EXTRACT(payload, '$._pwa.serviceWorkerHeuristic') = 'true' OR JSON_EXTRACT(payload, '$._pwa.serviceWorkers') != "[]")
  )
 OR JSON_EXTRACT(payload, '$._blinkFeatureFirstUsed.Features.ServiceWorkerControlledPage') IS NOT NULL
 OR JSON_EXTRACT(payload, '$._blinkFeatureFirstUsed.Features.990') IS NOT NULL
 OR IFNULL(JSON_EXTRACT(report, '$.audits.service-worker.score'),'0') = '1'
```

So This saves all the lighthouse service worker score, the ServiceWorkerControlledPage metric, and all our PWA custom metric JSON into a new table called `httparchive.scratchspace.serviceworkers_2021_07_01_mobile`, which allows us to run cheap queries while we investigate this further. For example:

```sql
#standardSQL
SELECT
  SUM(IF(lh_service_worker = '1',1,0)) AS lighthouse,
  SUM(ServiceWorkerControlledPage) AS blink,
  SUM(ServiceWorker) AS ServiceWorker,
  SUM(workboxInfo) AS workboxInfo,
  SUM(swEventListenersInfo) AS swEventListenersInfo,
  SUM(swMethodsInfo) AS swMethodsInfo,
  SUM(serviceWorkerHeuristic) AS serviceWorkerHeuristic
FROM
  `httparchive.scratchspace.serviceworkers_2021_07_01_mobile`
```
Which gives:

lighthouse | blink | ServiceWorker | workboxInfo | swEventListenersInfo | swMethodsInfo | serviceWorkerHeuristic
-- | -- | -- | -- | -- | -- | --
107,775 | 86,175 | 72,366 | 57,524 | 218,791 | 139,458 | 230,067

So it's the `swEventListenersInfo` and `swMethodsInfo` which are driving up the `serviceWorkerHeuristic` count.

If I look at individual pages for ones our heuristic thinks uses a service worker but the rest don't:

```sql
#standardSQL
SELECT
    url,
    JSON_EXTRACT(pwaCustomMetric, '$.serviceWorkers') AS ServiceWorker,
    JSON_EXTRACT(pwaCustomMetric, '$.swEventListenersInfo') AS swEventListenersInfo,
    JSON_EXTRACT(pwaCustomMetric, '$.swMethodsInfo') AS swMethodsInfo
FROM
  `httparchive.scratchspace.serviceworkers_2021_07_01_mobile`
WHERE
  lh_service_worker = '0'
  AND ServiceWorkerControlledPage = 0
  AND ServiceWorker = 0
  AND serviceWorkerHeuristic = 1
```

When I run a few of them through Chrome they do use a service worker.

The `serviceWorker` custom metric looks for an instance of `navigator.serviceWorker.register` and expects the provided file to also be in the URL (which we know doesn't always work for base URL) but misses things like this (as the argument is a variable and not the URL):

```js
const serviceWorkerPath = '/runtime-service-worker.js?v=2';
navigator.serviceWorker.register(serviceWorkerPath, { scope: './' })
```

So it will definitely undercount. I really think we should enhance this custom metric to look for any instance of `navigator.serviceWorker.register`, as well as an instance that matches a url fetched. In fact maybe we should even drop that `serviceWorker` object since it will only catch simple ones?

Lighthouse sometimes fails the audit as they don't register a `start_url` which explains the differences with those. And suggests lighthouse undercounts too.

Which leaves us with blink_features. [This seems to be the code for that](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/service_worker/service_worker_container.cc;l=494;drc=a7dbdae762701a9f07a2d8d0f2335c4e911ceba2) but don't know Chrome code well enough to understand why it might undercount.

Not sure if you have any insights here @demianrenzulli or any contacts within Google that could explain if it's too cautious?

I have however found some examples where the heuristic got it wrong:
- https://lostcoastoutpost.com/ seems to have the serviceWorker commented out
- https://www.tkvertriebs-auto.de/ uses firebase but doesn't register a service worker
- ...etc.
Both of these have a swEventListenersInfo but no swMethodsInfo so maybe we should exclude those from the heuristics going forward.

So I'm not sure if we should use `serviveWorker` or `serviceWorkerHeuristics`? In fact I'm tempted to use the later (or perhaps a combination of `serviceWorker and `swMethodsInfo` not being null?) as think the former is undercounting based on spot checks.

WDYT @demianrenzulli ?

Reply:
Resolving this as we've decided to use `serviceWorkerHeuristics`.